src/pipeline/src/etl/processor.rs

```diff
@@ -264,58 +267,86 @@ fn parse_processor(doc: &yaml_rust::Yaml) -> Result<ProcessorBuilders, String> {
         urlencoding::PROCESSOR_URL_ENCODING => {
             ProcessorBuilders::UrlEncoding(UrlEncodingProcessorBuilder::try_from(value)?)
         }
-        _ => return Err(format!("unsupported {} processor", str_key)),
+        _ => return UnsupportedProcessorSnafu { processor: str_key }.fail(),
     };
 
     Ok(processor)
 }
 
-pub(crate) fn yaml_string(v: &yaml_rust::Yaml, field: &str) -> Result<String, String> {
-    v.as_str()
-        .map(|s| s.to_string())
-        .ok_or(format!("'{field}' must be a string"))
+pub(crate) fn yaml_string(v: &yaml_rust::Yaml, field: &str) -> Result<String> {
+    v.as_str().map(|s| s.to_string()).ok_or(
+        FieldMustBeTypeSnafu {
+            field: field.to_string(),
+            ty: "string".to_string(),
+        }
+        .build(),
+    )
 }
 
-pub(crate) fn yaml_strings(v: &yaml_rust::Yaml, field: &str) -> Result<Vec<String>, String> {
+pub(crate) fn yaml_strings(v: &yaml_rust::Yaml, field: &str) -> Result<Vec<String>> {
     let vec = v
         .as_vec()
-        .ok_or(format!("'{field}' must be a list of strings",))?
+        .ok_or(
+            FieldMustBeTypeSnafu {
+                field: field.to_string(),
+                ty: "list of string".to_string(),
+            }
+            .build(),
+        )?
         .iter()
         .map(|v| v.as_str().unwrap_or_default().into())
         .collect();
     Ok(vec)
 }
 
-pub(crate) fn yaml_bool(v: &yaml_rust::Yaml, field: &str) -> Result<bool, String> {
-    v.as_bool().ok_or(format!("'{field}' must be a boolean"))
+pub(crate) fn yaml_bool(v: &yaml_rust::Yaml, field: &str) -> Result<bool> {
+    v.as_bool().ok_or(
+        FieldMustBeTypeSnafu {
+            field: field.to_string(),
+            ty: "boolean".to_string(),
+        }
+        .build(),
+    )
 }
 
-pub(crate) fn yaml_parse_string<T>(v: &yaml_rust::Yaml, field: &str) -> Result<T, String>
+pub(crate) fn yaml_parse_string<T>(v: &yaml_rust::Yaml, field: &str) -> Result<T>
 where
     T: std::str::FromStr,
-    T::Err: ToString,
+    T::Err: std::error::Error + Send + Sync + 'static,
 {
-    yaml_string(v, field)?
-        .parse::<T>()
-        .map_err(|e| e.to_string())
+    yaml_string(v, field)?.parse::<T>().map_err(|e| {
+        FailedParseFieldFromStringSnafu {
+            error: Box::new(e),
+            field: field.to_string(),
+        }
+        .build()
+    })
 }
 
-pub(crate) fn yaml_parse_strings<T>(v: &yaml_rust::Yaml, field: &str) -> Result<Vec<T>, String>
+pub(crate) fn yaml_parse_strings<T>(v: &yaml_rust::Yaml, field: &str) -> Result<Vec<T>>
 where
     T: std::str::FromStr,
-    T::Err: ToString,
+    T::Err: std::error::Error + Send + Sync + 'static,
 {
     yaml_strings(v, field).and_then(|v| {
         v.into_iter()
-            .map(|s| s.parse::<T>().map_err(|e| e.to_string()))
+            .map(|s| {
+                s.parse::<T>().map_err(|e| {
```

Comment:
I think there is a `context` method for this type of usage

```
s.parse::<T>().context(FailedParseFieldFromStringSnafu{ field: field.to_string() })
```

Reply:
We can't do that, because some errors from other third-party libraries can't be used as a source of Errors because of `stack_trace_debug`。